### ISSUE TITLE
ScanTask memory miscalculation causing OOM

### ISSUE THREAD

**Issue by sagiahrac** [Eventual-Inc/Daft#2383]
**Describe the bug**
ScanTask memory resources are set to the compressed size instead of the uncompressed byte size when working with row groups:
https://github.com/Eventual-Inc/Daft/blob/395ebe8f406488b11eb65d71c0ff01ba76a61d76/src/daft-scan/src/scan_task_iters.rs#L189

Similarly, when the number of Parquet files exceeds `parquet_split_row_groups_max_files`, the memory requirement is based on the compressed file size. Instead, the uncompressed size of the selected columns should be used, or the compressed size should be multiplied by an inflation factor (configurable or `parquet_inflation_factor`).

Finally, the system currently ignores already used resources when checking for used memory:
https://github.com/Eventual-Inc/Daft/blob/395ebe8f406488b11eb65d71c0ff01ba76a61d76/src/common/system-info/src/lib.rs#L38

I would expect the system to check the available memory at dispatch time:
https://docs.rs/sysinfo/latest/sysinfo/struct.System.html#method.available_memory

**To Reproduce**
env:
```
python==3.11.9daft==0.2.24
pandas==2.2.2
numpy==1.26.4
pytest==7.4.4
```

code:
```python
import os

import daft
import numpy as np
import pandas as pd
import pyarrow.parquet as pq
import pytest

PARQUET_PATH = "zeros.parquet"
N_ROW_GROUPS = 1
ROW_GROUP_SIZE = int(1e6)
N_ROWS = N_ROW_GROUPS * ROW_GROUP_SIZE


@pytest.fixture
def dummy_parquet(dst: str = PARQUET_PATH):
    pdf = pd.DataFrame({"A": np.zeros(N_ROWS, dtype=np.int64)})
    pdf.to_parquet(
        dst, compression="gzip", use_dictionary=False, row_group_size=ROW_GROUP_SIZE
    )
    yield dst
    os.remove(dst)


@pytest.fixture
def uncompressed_file_size(dummy_parquet: str):
    pf = pq.ParquetFile(dummy_parquet)
    metadata = pf.metadata
    n_row_groups = metadata.num_row_groups
    return sum(metadata.row_group(i).total_byte_size for i in range(n_row_groups))


def patch_pyrunner_memory(bytes_memory: int):
    pyrunner = daft.context.get_context()._get_runner()
    pyrunner.bytes_memory = bytes_memory


def test_lessthan_uncompressed_file_size(dummy_parquet, uncompressed_file_size):
    patch_pyrunner_memory(uncompressed_file_size // 2)

    df = daft.read_parquet(dummy_parquet)
    with pytest.raises(RuntimeError):  # because memory limit is less than uncompressed file size
        next(iter(df))
    # RuntimeError is not raised -> test fails.

```

**Expected behavior**

- When running the `test_lessthan_uncompressed_file_size` test, a `RuntimeError` should be raised because the available memory is set to less than half of the uncompressed file size. This will confirm that the system correctly accounts for the uncompressed size or applies an appropriate inflation factor when estimating memory requirements.

- In general, `SystemInfo.total_memory` will consider only available memory.

**Comment by GuyPozner** [comment]
Adding some more info:
1. Using the size on disk(compressed size) for each row group to estimate the scan task size in memory causes the PyRunner to underestimate the amount of memory needed for that scan task, causing the pyrunner to admit more tasks then the machine can actually fulfill.
2. Using SystemInfo.total_memory, overestimates the amount of free memory that the pyrunner have to admit tasks, thus admiting tasks that when materialize cause an OOM.
3. It all boils down to the following line: https://github.com/Eventual-Inc/Daft/blob/395ebe8f406488b11eb65d71c0ff01ba76a61d76/daft/runners/pyrunner.py#L312-L314

**Comment by jaychia** [comment]
Great points! We actually were just debugging this with a user last week.

> ScanTask memory resources are set to the compressed size instead of the uncompressed byte size when working with row groups

This definitely needs to be fixed, we're looking into it now.

**Comment by kevinzwang** [comment]
Hi @GuyPozner @sagiahrac, We just released v0.2.29 which uses an estimated in-memory size for both scan task merging and task scheduling, so you can now call `daft.set_execution_config(parquet_inflation_factor=...)` to prevent those OOMs. Let me know if you run into any problems with using that, otherwise feel free to close this issue!

**Comment by GuyPozner** [comment]
Thanks, we are testing it, and we will close it if it works.

**Comment by sagiahrac** [comment]
> Hi @GuyPozner @sagiahrac, We just released v0.2.29 which uses an estimated in-memory size for both scan task merging and task scheduling, so you can now call `daft.set_execution_config(parquet_inflation_factor=...)` to prevent those OOMs. Let me know if you run into any problems with using that, otherwise feel free to close this issue!

Thanks! We tried that, and it worked great. Closing this issue.
